pkg/cloud/powervs.go

```
@@ -155,22 +160,48 @@ func (p *powerVSCloud) CreateDisk(volumeName string, diskOptions *DiskOptions) (
+	if _, ok := p.diskOpMap[volumeName]; !ok {
+		// Mark as already op-ed.
+		p.diskOpMap[volumeName] = struct{}{}
+		_, err := p.volClient.CreateVolume(dataVolume)
+		if err != nil {
+			// worst case, the pod restarted and lost the map. Handle such case with finding the available volume on cloud.
+			// based on the diskOp Map, there is an entry, so check if it's available and proceed to check for volume state
+			if strings.Contains(strings.ToLower(err.Error()), ErrVolumeNameAlreadyExists.Error()) {
+				return p.ensureVolumeAvailable(volumeName)
+			}
+			// Some other error apart from the creation of volume with conflicting names.
+			return nil, err
+		}
 	}
+	return p.ensureVolumeAvailable(volumeName)
+}
 
-	err = p.WaitForVolumeState(*v.VolumeID, VolumeAvailableState)
-	if err != nil {
-		return nil, err
+func (p *powerVSCloud) ensureVolumeAvailable(volumeName string) (*Disk, error) {
+	disk, _ := p.GetDiskByName(volumeName)
+	if disk.State != VolumeAvailableState {
+		err := p.WaitForVolumeState(disk.VolumeID, VolumeAvailableState)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Disk exists, but not in required state. Current:%s Required:%s", disk.State, VolumeAvailableState)
+		}
 	}
-
-	return &Disk{CapacityGiB: capacityGiB, VolumeID: *v.VolumeID, DiskType: v.DiskType, WWN: strings.ToLower(v.Wwn)}, nil
+	delete(p.diskOpMap, volumeName)
+	return &Disk{CapacityGiB: disk.CapacityGiB, VolumeID: disk.VolumeID, DiskType: disk.DiskType, WWN: strings.ToLower(disk.WWN)}, nil
 }
-
 func (p *powerVSCloud) DeleteDisk(volumeID string) (err error) {
-	return p.volClient.DeleteVolume(volumeID)
+	klog.V(4).Infof("Deleting Disk with ID: %s", volumeID)
+	start := time.Now()
+	err = p.volClient.DeleteVolume(volumeID)
+	if err != nil {
+		if strings.Contains(strings.ToLower(err.Error()), ErrVolumeNotFound.Error()) {
+			klog.Warningf("Volume %s not found, assuming deleted", volumeID)
+			return nil
+		}
+		return err
+	}
```

If the volume not found error is not handled:
```
Events:
  Type     Reason              Age                   From                                                                                             Message
  ----     ------              ----                  ----                                                                                             -------
...
  Warning  VolumeFailedDelete  58s (x3 over 3m10s)   powervs.csi.ibm.com_powervs-csi-controller-9f7dfd45b-4vnps_9219da62-c812-4d8f-9868-c84afca35d70  rpc error: code = Internal desc = Could not delete volume ID "44ccac77-b0ad-4880-a7d1-c8880151d07b": failed to perform Delete Volume Operation for volume 44ccac77-b0ad-4880-a7d1-c8880151d07b with error [DELETE /pcloud/v1/cloud-instances/{cloud_instance_id}/volumes/{volume_id}][404] pcloudCloudinstancesVolumesDeleteNotFound {"description":"volume does not exist. ID: 44ccac77-b0ad-4880-a7d1-c8880151d07b","error":"volume not found"}
```
With the check, #754 can be mitigated.